### ISSUE TITLE
Different best pipelines when random seed is fixed

### ISSUE THREAD

**Issue by JessyD** [EpistasisLab/tpot#813]
Every run of the TPOT model gives me a different best pipeline even when the random seed is fixed. 

## Process to reproduce the issue

I am using TPOT==0.9.5 and scikit-learn==0.19.2 and running the code below multiple times:

```
from tpot import TPOTRegressor
from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split

random_seed = 42
housing = load_boston()
X_train, X_test, y_train, y_test = \
train_test_split(housing.data, housing.target, train_size=0.75, test_size=0.25,
                random_state=random_seed)
scoring = 'neg_mean_absolute_error'

tpot = TPOTRegressor(generations=30,
                     population_size=50,
                     verbosity=2,
                     random_state=random_seed,
                     config_dict='TPOT light',
                     scoring=scoring
                     )
tpot.fit(X_train, y_train)
print('Test score using optimal model: %f ' %tpot.score(X_test, y_test))
tpot.export('tpot_boston_pipeline.py')
```

## Expected result
As the random seed is fixed, I would expect the best suggested pipeline to be the same throughout the different runs.

## Current result
To illustrate the models below are the best models after running the code 3 times.
```
Best pipeline: DecisionTreeRegressor(LinearSVR(DecisionTreeRegressor(RobustScaler(StandardScaler(ElasticNetCV(input_matrix, l1_ratio=0.1, tol=0.001))), max_depth=5, min_samples_leaf=2, min_samples_split=11), C=20.0, dual=True, epsilon=1.0, loss=squared_epsilon_insensitive, tol=0.01), max_depth=6, min_samples_leaf=13, min_samples_split=17)
Test score using optimal model: -2.492078
```

```
Best pipeline: RidgeCV(ElasticNetCV(MaxAbsScaler(DecisionTreeRegressor(ElasticNetCV(input_matrix, l1_ratio=0.6000000000000001, tol=0.01), max_depth=5, min_samples_leaf=2, min_samples_split=12)), l1_ratio=0.7000000000000001, tol=0.0001))
Test score using optimal model: -2.473574
```
```
Best pipeline: LassoLarsCV(MinMaxScaler(CombineDFs(Normalizer(DecisionTreeRegressor(input_matrix, max_depth=5, min_samples_leaf=9, min_samples_split=18), norm=max), input_matrix)), normalize=True)
Test score using optimal model: -2.563920
```

**Comment by weixuanfu** [comment]
I just tested the demo in my environment, the best pipeline is reproduced with fixed random seed. Could you please let me know more details of the environment for reproducing this issue?
```
In [1]: from tpot import TPOTRegressor
   ...: from sklearn.datasets import load_boston
   ...: from sklearn.model_selection import train_test_split
   ...:
   ...: random_seed = 42
   ...: housing = load_boston()
   ...: X_train, X_test, y_train, y_test = \
   ...: train_test_split(housing.data, housing.target, train_size=0.75, test_size=0.25,
   ...:                 random_state=random_seed)
   ...: scoring = 'neg_mean_absolute_error'
   ...: for _ in range(3):
   ...:     tpot = TPOTRegressor(generations=30,
   ...:                      population_size=50,
   ...:                      verbosity=2,
   ...:                      random_state=random_seed,
   ...:                      config_dict='TPOT light',
   ...:                      scoring=scoring
   ...:                      )
   ...:     tpot.fit(X_train, y_train)
   ...:     print('Test score using optimal model: %f ' %tpot.score(X_test, y_test))
   ...:
Generation 1 - Current best internal CV score: -2.8928189002582902
Generation 2 - Current best internal CV score: -2.8928189002582902
Generation 3 - Current best internal CV score: -2.8928189002582902
Generation 4 - Current best internal CV score: -2.8928189002582902
Generation 5 - Current best internal CV score: -2.8928189002582902
Generation 6 - Current best internal CV score: -2.8928189002582902
Generation 7 - Current best internal CV score: -2.7588987549669652
Generation 8 - Current best internal CV score: -2.7588987549669652
Generation 9 - Current best internal CV score: -2.754552943044516
Generation 10 - Current best internal CV score: -2.754552943044516
Generation 11 - Current best internal CV score: -2.754552943044516
Generation 12 - Current best internal CV score: -2.713925278558578
Generation 13 - Current best internal CV score: -2.713925278558578
Generation 14 - Current best internal CV score: -2.713925278558578
Generation 15 - Current best internal CV score: -2.713925278558578
Generation 16 - Current best internal CV score: -2.713925278558578
Generation 17 - Current best internal CV score: -2.713925278558578
Generation 18 - Current best internal CV score: -2.713925278558578
Generation 19 - Current best internal CV score: -2.688096725715785
Generation 20 - Current best internal CV score: -2.688096725715785
Generation 21 - Current best internal CV score: -2.688096725715785
Generation 22 - Current best internal CV score: -2.688096725715785
Generation 23 - Current best internal CV score: -2.688096725715785
Generation 24 - Current best internal CV score: -2.6739259845515706
Generation 25 - Current best internal CV score: -2.644503760377815
Generation 26 - Current best internal CV score: -2.644503760377815
Generation 27 - Current best internal CV score: -2.644503760377815
Generation 28 - Current best internal CV score: -2.644419942090466
Generation 29 - Current best internal CV score: -2.644419942090466
Generation 30 - Current best internal CV score: -2.617657608544968

Best pipeline: ElasticNetCV(MaxAbsScaler(RobustScaler(DecisionTreeRegressor(SelectFwe(ElasticNetCV(LinearSVR(MaxAbsScaler(DecisionTreeRegressor(ElasticNetCV(input_matrix, l1_ratio=0.5, tol=0.01), max_depth=5, min_samples_leaf=2, min_samples_split=12)), C=20.0, dual=True, epsilon=0.1, loss=epsilon_insensitive, tol=0.0001), l1_ratio=0.5, tol=0.01), alpha=0.024), max_depth=5, min_samples_leaf=2, min_samples_split=12))), l1_ratio=0.1, tol=0.001)
Test score using optimal model: -2.321129
Generation 1 - Current best internal CV score: -2.8928189002582902
Generation 2 - Current best internal CV score: -2.8928189002582902
Generation 3 - Current best internal CV score: -2.8928189002582902
Generation 4 - Current best internal CV score: -2.8928189002582902
Generation 5 - Current best internal CV score: -2.8928189002582902
Generation 6 - Current best internal CV score: -2.8928189002582902
Generation 7 - Current best internal CV score: -2.7588987549669652
Generation 8 - Current best internal CV score: -2.7588987549669652
Generation 9 - Current best internal CV score: -2.754552943044516
Generation 10 - Current best internal CV score: -2.754552943044516
Generation 11 - Current best internal CV score: -2.754552943044516
Generation 12 - Current best internal CV score: -2.713925278558578
Generation 13 - Current best internal CV score: -2.713925278558578
Generation 14 - Current best internal CV score: -2.713925278558578
Generation 15 - Current best internal CV score: -2.713925278558578
Generation 16 - Current best internal CV score: -2.713925278558578
Generation 17 - Current best internal CV score: -2.713925278558578
Generation 18 - Current best internal CV score: -2.713925278558578
Generation 19 - Current best internal CV score: -2.688096725715785
Generation 20 - Current best internal CV score: -2.688096725715785
Generation 21 - Current best internal CV score: -2.688096725715785
Generation 22 - Current best internal CV score: -2.688096725715785
Generation 23 - Current best internal CV score: -2.688096725715785
Generation 24 - Current best internal CV score: -2.6739259845515706
Generation 25 - Current best internal CV score: -2.644503760377815
Generation 26 - Current best internal CV score: -2.644503760377815
Generation 27 - Current best internal CV score: -2.644503760377815
Generation 28 - Current best internal CV score: -2.644419942090466
Generation 29 - Current best internal CV score: -2.644419942090466
Generation 30 - Current best internal CV score: -2.617657608544968

Best pipeline: ElasticNetCV(MaxAbsScaler(RobustScaler(DecisionTreeRegressor(SelectFwe(ElasticNetCV(LinearSVR(MaxAbsScaler(DecisionTreeRegressor(ElasticNetCV(input_matrix, l1_ratio=0.5, tol=0.01), max_depth=5, min_samples_leaf=2, min_samples_split=12)), C=20.0, dual=True, epsilon=0.1, loss=epsilon_insensitive, tol=0.0001), l1_ratio=0.5, tol=0.01), alpha=0.024), max_depth=5, min_samples_leaf=2, min_samples_split=12))), l1_ratio=0.1, tol=0.001)
Test score using optimal model: -2.321129
Generation 1 - Current best internal CV score: -2.8928189002582902
Generation 2 - Current best internal CV score: -2.8928189002582902
Generation 3 - Current best internal CV score: -2.8928189002582902
Generation 4 - Current best internal CV score: -2.8928189002582902
Generation 5 - Current best internal CV score: -2.8928189002582902
Generation 6 - Current best internal CV score: -2.8928189002582902
Generation 7 - Current best internal CV score: -2.7588987549669652
Generation 8 - Current best internal CV score: -2.7588987549669652
Generation 9 - Current best internal CV score: -2.754552943044516
Generation 10 - Current best internal CV score: -2.754552943044516
Generation 11 - Current best internal CV score: -2.754552943044516
Generation 12 - Current best internal CV score: -2.713925278558578
Generation 13 - Current best internal CV score: -2.713925278558578
Generation 14 - Current best internal CV score: -2.713925278558578
Generation 15 - Current best internal CV score: -2.713925278558578
Generation 16 - Current best internal CV score: -2.713925278558578
Generation 17 - Current best internal CV score: -2.713925278558578
Generation 18 - Current best internal CV score: -2.713925278558578
Generation 19 - Current best internal CV score: -2.688096725715785
Generation 20 - Current best internal CV score: -2.688096725715785
Generation 21 - Current best internal CV score: -2.688096725715785
Generation 22 - Current best internal CV score: -2.688096725715785
Generation 23 - Current best internal CV score: -2.688096725715785
Generation 24 - Current best internal CV score: -2.6739259845515706
Generation 25 - Current best internal CV score: -2.644503760377815
Generation 26 - Current best internal CV score: -2.644503760377815
Generation 27 - Current best internal CV score: -2.644503760377815
Generation 28 - Current best internal CV score: -2.644419942090466
Generation 29 - Current best internal CV score: -2.644419942090466
Generation 30 - Current best internal CV score: -2.617657608544968

Best pipeline: ElasticNetCV(MaxAbsScaler(RobustScaler(DecisionTreeRegressor(SelectFwe(ElasticNetCV(LinearSVR(MaxAbsScaler(DecisionTreeRegressor(ElasticNetCV(input_matrix, l1_ratio=0.5, tol=0.01), max_depth=5, min_samples_leaf=2, min_samples_split=12)), C=20.0, dual=True, epsilon=0.1, loss=epsilon_insensitive, tol=0.0001), l1_ratio=0.5, tol=0.01), alpha=0.024), max_depth=5, min_samples_leaf=2, min_samples_split=12))), l1_ratio=0.1, tol=0.001)
Test score using optimal model: -2.321129
```

**Comment by JessyD** [comment]
I ran the same script with the same environment in a singularity container (running ubuntu 18.04) and also obtained consistent results. 

However, if I try it on an OSX 10.12.6, I get different results for every run. Apart from the underlying operating system, I am using the same package versions in both environments. 

My current environment looks like this:
```
pip3 --no-cache-dir install \
         numpy==1.15.1 \
         pandas==0.23.4 \
         scipy==1.1.0 \
         scikit-learn==0.19.2 \
         nibabel==2.3.0 \
         nilearn==0.4.2 \
         matplotlib==3.0.0 \
         ipython==7.0.1 \
         jupyter notebook==5.6.0 \
         TPOT==0.9.5 \
         bokeh==0.13.0 \
         dask==0.19.4 \
         distributed==1.23.3\
         dask-ml==0.10.0\
         dask-glm==0.1.0\
         joblib==0.12.5\
         tornado==5.1.1\
         seaborn==0.9.0"
```
I will just use the singularity environment in the future.
